### DISCUSSION TITLE
Tables and spaces - and some problems

### DISCUSSION THREAD

**Discussion by Nick222** [zim-desktop-wiki/zim-desktop-wiki#1827] [General]
Unfortunately, the use of multiple spaces in tables to align them creates certain problems.

For example, when trying to find and remove double spaces, when searching for single words, etc.

Can you replace spaces in tables with a different alignment method?

Thank you :)

**Comment by introt** [comment]
Do you have any suggestions? I'm not seeing huge gains from any alternatives, while all of them would break backwards compatibility and most wouldn't be hand-editable.

> when trying to find and remove double spaces

The parser doesn't care too much about alignment, it's the `|`s where the magic is at. Also, Zim's own find-and-replace should ignore formatting, so you can use that to be extra safe.

> when searching for single words

Can you elaborate on how the multiple spaces create problems here? Doesn't sound like anything a little regex couldn't handle ;)

**Comment by Nick222** [comment]
I don't think you need to force the average user Zim to learn regular expressions.
Do not forget that the simplicity of the format of this program makes it very easy (technically and psychologically) to edit its files using third-party software - which potentially leads to conflicts between different edits.

In addition, it is very difficult to work with line breaks in table cells, which, in my opinion, requires its own solution.
In general, as a user, the implementation of tables in Zim gives me a feeling of ill-conceived and unfinished business.

I think that the source text of the note needs a label for the beginning of the table and separators of table cells, and formatting (including alignment) should be done by the program itself.

If you want, we can start a discussion about the use of tables in Zim, especially since there are already a lot of complaints about this option ...

**Reply by introt** [reply]
I've converted this issue into a discussion as there are, well, many things to discuss here. Disclaimer: opinions below are mine may not align with those of Jaap.

I agree with you about the regular expressions - but if the issue is with searching for whole words, we should improve the in-application search instead of table formatting.

Conflicts happen when the same file is opened simultaneously by multiple applications, which can't be alleviated via formatting - do you mean something else here? Could you give some examples?

What is your current use case where the current formatting table formatting gets in the way like? Don't get me wrong, I'm also not a fan of the current table plugin, but I think its faults lie in the editor, not the syntax, as I will try to demonstrate below:

"a label for the beginning of the table": 
```
| column |titles|
|-| --|
```
"separators of table cells": `|` and `\n` (line break, not literal)
```
| hello |there |
|general|kenobi|
```

"formatting (including alignment) should be done by the program itself": this is how it works already (the parser doesn't care about the number or lack of spaces):
![image](https://user-images.githubusercontent.com/26069787/146815856-40171f8e-43ae-472f-84a3-55c6f8452196.png)
![image](https://user-images.githubusercontent.com/26069787/146815911-9f9f938a-858d-4ae3-b750-bb1c65f646f8.png)

If you modify the source text file manually, you need to interact with the table to force a rewrite of the source; as you can see, the source is both aligned and includes the column alignment (`:`):

```
| column  | titles |
|:--------|-------:|
| hello   |  there |
| general | kenobi |
```

You seem to be asking for HTML-style table syntax; that isn't in line with the wiki style nor is it human readable (in the broadest sense). I'd rather have a general HTML-embedding plugin and keep the current table plugin as-is, simple and comparable to the MarkDown equivalent.

Using literal `\n` for line breaks in the source keeps rows contained in a single line and thus `grep`able. Requiring the user to type `\n` in Zim is an editor limitation fixable by implementing a `Ctrl`/`Shift` + `Enter` shortcut as many table editors do.

Totally okay with having `{{image.png}}` support and/or TiddlyWiki-style `{{transclusion}}` stuff inside those cells though, that's right up my alley; the latter would also side-step the line break (and likely the "single-word") problem.

Thanks for bringing up your issues, we can make Zim better together!
